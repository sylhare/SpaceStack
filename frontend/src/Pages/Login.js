import React, {Component} from 'react';
import {connect} from 'react-redux'
import {Redirect} from 'react-router-dom';
import {login} from '../Actions/auth';

let createHandlers = function (dispatch) {
  let loginOnSubmit = function (username, password) {
    dispatch(login(username, password))
      .catch(() => {
      });
  };

  return {loginOnSubmit};
};


class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      username: '',
      password: '',
      errorUsername: '',
      errorPassword: '',
    };
    this.handlers = createHandlers(this.props.dispatch);
  }

  handleSubmit = (evt) => {
    evt.preventDefault();
    this.setState({errorUsername: this.state.username ? '' : 'Username is required'});
    this.setState({errorPassword: this.state.password ? '' : 'Password is required'});


    if (this.state.username && this.state.password) {
      this.handlers.loginOnSubmit(this.state.username, this.state.password)
    }
  };

  handleUserChange = (evt) => {
    this.setState({
      username: evt.target.value,
      errorUsername: ''
    });
  };

  handlePassChange = (evt) => {
    this.setState({
      password: evt.target.value,
      errorPassword: ''
    });
  };

  render() {
    if (this.props.isLoggedIn) {
      return <Redirect to='/'/>;
    } else {
      return (
        <div className='content'>
          <div className='login'>
            <form onSubmit={this.handleSubmit}>
              <label className='login-input'><b>Username</b></label>
              <input type='text'
                     data-test='username'
                     className='login-input'
                     placeholder='Enter Username'
                     value={this.state.username} onChange={this.handleUserChange}/>
              {this.state.errorUsername && <span className='error'>{this.state.errorUsername}</span>}
              <label className='login-input'><b>Password</b></label>
              <input type='password'
                     className='login-input'
                     data-test='password'
                     placeholder='Enter Password'
                     value={this.state.password} onChange={this.handlePassChange}/>
              {this.state.errorPassword && <span className='error'>{this.state.errorPassword}</span>}
              <div className='flex-container'>
                <button type='submit'
                        aria-label='login'
                        className='login-button'
                        data-test='submit'>Log in
                </button>
              </div>
            </form>
          </div>
        </div>
      );
    }
  }
}

const mapStateToProps = state => {
  return {
    isLoggedIn: state.authReducer.isLoggedIn,
  };
};


export default connect(mapStateToProps)(Login);
