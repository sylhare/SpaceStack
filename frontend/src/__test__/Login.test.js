import React from 'react'
import {rest} from 'msw'
import {setupServer} from 'msw/node'
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'
import Login from "../Pages/Login";

import configureStore from 'redux-mock-store';
import {Provider} from "react-redux";

const mockStore = configureStore([]);


const server = setupServer(
  rest.get('/v1/login', (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({token: 'this is the token'}))
  }),
);

let store;
let component;

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
beforeEach(() => {
  store = mockStore({
    authReducer: {
      isLoggedIn: false,
    }
  });

  store.dispatch = jest.fn().mockImplementation(() => Promise.resolve());

  component = renderer.create(
    <Provider store={store}>
      <Login dispatch={store.dispatch}/>
    </Provider>
  );
});

it('should render with given state from Redux store', () => {
  expect(component.toJSON()).toMatchSnapshot();
});

it('should login with username password', () => {
  renderer.act(() => {
    component.root.findByProps({'data-test': 'username'}).props.onChange({target: {value: "username"}});
    component.root.findByProps({'data-test': 'password'}).props.onChange({target: {value: "password"}});
  });

  renderer.act(() => {
    component.root.findByType('form').props.onSubmit({preventDefault: jest.fn()});
  });

  expect(store.dispatch).toHaveBeenCalledTimes(1);
});

it('should not login if username missing', () => {
  renderer.act(() => {
    component.root.findByProps({'data-test': 'password'}).props.onChange({target: {value: "password"}});
  });

  renderer.act(() => {
    component.root.findByType('form').props.onSubmit({preventDefault: jest.fn()});
  });

  expect(store.dispatch).toHaveBeenCalledTimes(0);
});

it('should not login if password missing', () => {
  renderer.act(() => {
    component.root.findByProps({'data-test': 'username'}).props.onChange({target: {value: "username"}});
  });

  renderer.act(() => {
    component.root.findByType('form').props.onSubmit({preventDefault: jest.fn()});
  });

  expect(store.dispatch).toHaveBeenCalledTimes(0);
});
