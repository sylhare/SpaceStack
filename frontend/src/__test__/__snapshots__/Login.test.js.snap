// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with given state from Redux store 1`] = `
<div
  className="content"
>
  <div
    className="login"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        className="login-input"
      >
        <b>
          Username
        </b>
      </label>
      <input
        className="login-input"
        data-test="username"
        onChange={[Function]}
        placeholder="Enter Username"
        type="text"
        value=""
      />
      
      <label
        className="login-input"
      >
        <b>
          Password
        </b>
      </label>
      <input
        className="login-input"
        data-test="password"
        onChange={[Function]}
        placeholder="Enter Password"
        type="password"
        value=""
      />
      
      <div
        className="flex-container"
      >
        <button
          aria-label="login"
          className="login-button"
          data-test="submit"
          type="submit"
        >
          Log in
        </button>
      </div>
    </form>
  </div>
</div>
`;
